src/deviation.jl

```diff
@@ -9,13 +9,10 @@ Count the number of indices at which the elements of the arrays
 `a` and `b` are equal.
 """
 function counteq(a::AbstractArray, b::AbstractArray)
-    n = length(a)
-    length(b) == n || throw(DimensionMismatch("Inconsistent lengths."))
+    length(a) == length(b) || throw(DimensionMismatch("Inconsistent lengths."))
     c = 0
-    for i = 1:n
-        @inbounds if a[i] == b[i]
-            c += 1
-        end
+    for (ai, bi) = zip(a, b)
```

Comment:
Using `zip` here is incorrect given the contract of `counteq` (same comments below).

The point is that the contract states:

> Count the number of indices

and `zip` will ignore indices but instead lead to comparing values using their iteration order.

Reply:
IMO that's consistent with the current behaviour though, regardless of the docstring. The current implementation already only cares about whether the first, second, third, etc elements match but doesn't care about shapes or cartesian indices of the compared arrays.

Reply:
Another nice consequence of using `zip` is that it is consistent with how Distances handles input arrays.

Reply:
I agree with @bkamins that the docstring implies that values at the same indices are compared. So we need to check that indices are the same (or that starting indices are equal in addition to the current check that length are equal). We can keep a tolerance when mixing vectors and matrices by only check that linear indices are equal to avoid breakage.

What do you mean about Distances? Doesn't it check that axes are the same or that at least linear indices are the same?

Reply:
Since the package does neither support nor test arrays with non-standard or non-linear indices, my interpretation is that the docstring just refers to iteration order but it was not intended to come up with a more general design decision. Maybe there's some information in the original PR.

Reply:
Yes we're free to reinterpret and adapt the docstring as we deem appropriate. But when passing arrays with different axes, it would find it dangerous to silently discard indices. If people use `OffsetArray`s there must be a reason. That's also a safer approach as if people find it too inconvenient, we can relax this requirement and allow mismatched indices. OTC if we allow them now we won't be able to change this later.

Reply:
I think it is clear from the source codes that the original design did not take into account non-1-based indices. Then - naturally - docstrings do not reflect this case.

I think we need to make a general decision for the whole package:
- if we allow non-1-based indexing (I feel from the discussions that we want to allow them)
- if yes - how they should be treated (and this should be made consistent across all methods in the package)
- what is our position about mixing arrays of different dimensionalities (this point is related as higher-dimensional arrays usually support linear indexing, which is 1-based) and when we want to accept only vectors.

After these decisions are made and documented the respective PRs should be done to reflect them. Otherwise we risk a situation when different methods in the package will take different assumptions.

Reply:
A simple solution could be to adopt the semantics of `broadcast`, as discussed above at https://github.com/JuliaStats/StatsBase.jl/pull/790#discussion_r874217136. This would avoid forcing users to understand yet another rule and it would be easier to implement for us (in practice we could use a different implementation under the hood if needed for performance).